FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/DemoTasks/FleetProvisioningDemoExample.c

```diff
@@ -0,0 +1,831 @@
+ * Provisioning feature.
+ *
+ * The Fleet Provisioning library provides macros and helper functions for
+ * assembling MQTT topics strings, and for determining whether an incoming MQTT
+ * message is related to the Fleet Provisioning API of AWS IoT Core. The Fleet
+ * Provisioning library does not depend on any particular MQTT library,
+ * therefore the functionality for MQTT operations is placed in another file
+ * (mqtt_operations.c). This demo uses the coreMQTT library. If needed,
+ * mqtt_operations.c can be modified to replace coreMQTT with another MQTT
+ * library. This demo requires using the AWS IoT Core broker as Fleet
+ * Provisioning is an AWS IoT Core feature.
+ *
+ * This demo provisions a device certificate using the provisioning by claim
+ * workflow with a Certificate Signing Request (CSR). The demo connects to AWS
+ * IoT Core using provided claim credentials (whose certificate needs to be
+ * registered with IoT Core before running this demo), subscribes to the
+ * CreateCertificateFromCsr topics, and obtains a certificate. It then
+ * subscribes to the RegisterThing topics and activates the certificate and
+ * obtains a Thing using the provisioning template. Finally, it reconnects to
+ * AWS IoT Core using the new credentials.
+ */
+
+/* Standard includes. */
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+
+/* Demo Config */
+#include "demo_config.h"
+
+/* mbedTLS include for configuring threading functions */
+#include "mbedtls/threading.h"
+#include "threading_alt.h"
+
+/* TinyCBOR library for CBOR encoding and decoding operations. */
+#include "cbor.h"
+
+/* corePKCS11 includes. */
+#include "core_pkcs11.h"
+#include "core_pkcs11_config.h"
+
+/* AWS IoT Fleet Provisioning Library. */
+#include "fleet_provisioning.h"
+
+/* Demo includes. */
+#include "mqtt_operations.h"
+#include "pkcs11_operations.h"
+#include "fleet_provisioning_serializer.h"
+#include "using_mbedtls_pkcs11.h"
+
+/**
+ * These configurations are required. Throw compilation error if it is not
+ * defined.
+ */
+#ifndef democonfigPROVISIONING_TEMPLATE_NAME
+    #error "Please define democonfigPROVISIONING_TEMPLATE_NAME to the template name registered with AWS IoT Core in demo_config.h."
+#endif
+#ifndef democonfigDEVICE_SERIAL_NUMBER
+    #error "Please define a serial number (democonfigDEVICE_SERIAL_NUMBER) in demo_config.h."
+#endif
+#ifndef democonfigROOT_CA_PEM
+    #error "Please define Root CA certificate of the MQTT broker(democonfigROOT_CA_PEM) in demo_config.h."
+#endif
+
+/**
+ * @brief The length of #democonfigPROVISIONING_TEMPLATE_NAME.
+ */
+#define fpdemoPROVISIONING_TEMPLATE_NAME_LENGTH    ( ( uint16_t ) ( sizeof( democonfigPROVISIONING_TEMPLATE_NAME ) - 1 ) )
+
+/**
+ * @brief The length of #democonfigDEVICE_SERIAL_NUMBER.
+ */
+#define fpdemoDEVICE_SERIAL_NUMBER_LENGTH          ( ( uint16_t ) ( sizeof( democonfigDEVICE_SERIAL_NUMBER ) - 1 ) )
```

Comment:
Is this the same as democonfigCLIENT_IDENTIFIER_LENGTH?

Reply:
No, the serial number and client identifiers are separate fields used for different purposes and most likely have separate lengths.

Reply:
Addressed changes to single unique identifier in this commit: https://github.com/FreeRTOS/FreeRTOS/pull/724/commits/518523e2fc48ecf6046901cc758eebd07fe274c0